File: hiroo-py/hiroo/http.py
"""Thin async HTTP wrapper over aiohttp for the HiRoo REST API."""
from __future__ import annotations

import asyncio
from typing import Any, List, Optional

import aiohttp

from .errors import HTTPError


class HTTPClient:
    def __init__(self, token: str, base_url: str):
        # Tokens are stored verbatim. For bots, `Authorization: Bot xxx` is
        # accepted by our API; for OAuth2 users, `Authorization: Bearer xxx`.
        self.token = token
        self.base_url = base_url.rstrip("/")
        self._session: Optional[aiohttp.ClientSession] = None

    @property
    def auth_header(self) -> str:
        if self.token.lower().startswith(("bot ", "bearer ")):
            return self.token
        return f"Bot {self.token}"

    async def _ensure(self) -> aiohttp.ClientSession:
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession(
                headers={"Authorization": self.auth_header, "User-Agent": "hiroo.py/0.1"}
            )
        return self._session

    async def close(self) -> None:
        if self._session and not self._session.closed:
            await self._session.close()

    async def request(self, method: str, path: str, *, json: Any = None, params: dict | None = None) -> Any:
        s = await self._ensure()
        url = self.base_url + path
        async with s.request(method, url, json=json, params=params) as r:
            if r.status == 204:
                return None
            try:
                data = await r.json()
            except aiohttp.ContentTypeError:
                data = await r.text()
            if r.status >= 400:
                raise HTTPError(r.status, data)
            return data

    # ── Convenience ────────────────────────────────────────────────────

    async def send_message(self, *, channel_id: Optional[str], dm_id: Optional[str], content: str = "",
                           embeds: list | None = None, components: list | None = None,
                           reply_to_id: Optional[str] = None) -> dict:
        body: dict = {"content": content}
        if embeds:
            body["embeds"] = [e.to_dict() if hasattr(e, "to_dict") else e for e in embeds]
        if components:
            body["components"] = [c.to_dict() if hasattr(c, "to_dict") else c for c in components]
        if reply_to_id:
            body["reply_to_id"] = reply_to_id
        if channel_id:
            return await self.request("POST", f"/api/channels/{channel_id}/messages", json=body)
        if dm_id:
            return await self.request("POST", f"/api/dms/{dm_id}/messages", json=body)
        raise ValueError("Either channel_id or dm_id required")

    async def register_command(self, application_id: str, *, type: str, name: str, description: str,
                               options: list | None = None, guild_id: str | None = None) -> dict:
        body = {
            "type": type, "name": name, "description": description,
            "options": options or [], "guild_id": guild_id,
        }
        return await self.request(
            "POST", f"/api/commands/applications/{application_id}/commands", json=body,
        )

    # ── Guild helpers ──────────────────────────────────────────────────

    async def get_guild_channels(self, guild_id: str) -> list[dict]:
        return await self.request("GET", f"/api/servers/{guild_id}/channels")

    async def create_channel(
        self,
        guild_id: str,
        name: str,
        *,
        type: str = "text",
        topic: str | None = None,
        is_private: bool = False,
        parent_id: str | None = None,
        position: int = 0,
    ) -> dict:
        """Create a channel in a server. `parent_id` points to a category
        channel to nest the new channel inside. Valid `type` values:
        `text`, `voice`, `announcement`, `category`, `forum`. Requires the
        bot to have MANAGE_CHANNELS in the guild."""
        body = {
            "name": name, "type": type, "is_private": is_private, "position": position,
        }
        if topic is not None:
            body["topic"] = topic
        if parent_id is not None:
            body["parent_id"] = parent_id
        return await self.request("POST", f"/api/servers/{guild_id}/channels", json=body)

    async def update_channel(self, guild_id: str, channel_id: str, **fields) -> dict:
        return await self.request(
            "PATCH", f"/api/servers/{guild_id}/channels/{channel_id}", json=fields,
        )

    async def delete_channel(self, guild_id: str, channel_id: str) -> None:
        await self.request("DELETE", f"/api/servers/{guild_id}/channels/{channel_id}")

    async def get_guilds(self) -> list[dict]:
        return await self.request("GET", "/api/servers")

    # ── Messages ───────────────────────────────────────────────────────

    async def edit_message(self, channel_id: str, message_id: str, content: str) -> dict:
        return await self.request(
            "PATCH", f"/api/channels/{channel_id}/messages/{message_id}",
            json={"content": content},
        )

    async def delete_message(self, channel_id: str, message_id: str) -> None:
        await self.request("DELETE", f"/api/channels/{channel_id}/messages/{message_id}")

    async def edit_dm_message(self, dm_id: str, message_id: str, content: str) -> dict:
        return await self.request(
            "PATCH", f"/api/dms/{dm_id}/messages/{message_id}",
            json={"content": content},
        )

    async def delete_dm_message(self, dm_id: str, message_id: str) -> None:
        await self.request("DELETE", f"/api/dms/{dm_id}/messages/{message_id}")

    async def add_reaction(self, channel_id: str, message_id: str, emoji: str) -> None:
        await self.request(
            "POST", f"/api/channels/{channel_id}/messages/{message_id}/reactions",
            json={"emoji": emoji},
        )

    async def remove_reaction(self, channel_id: str, message_id: str, emoji: str) -> None:
        await self.request(
            "DELETE", f"/api/channels/{channel_id}/messages/{message_id}/reactions",
            params={"emoji": emoji},
        )

    async def pin_message(self, channel_id: str, message_id: str) -> dict:
        return await self.request(
            "PUT", f"/api/channels/{channel_id}/messages/{message_id}/pin",
        )

    async def unpin_message(self, channel_id: str, message_id: str) -> dict:
        return await self.request(
            "DELETE", f"/api/channels/{channel_id}/messages/{message_id}/pin",
        )

    async def list_pinned(self, channel_id: str) -> list[dict]:
        return await self.request("GET", f"/api/channels/{channel_id}/messages/pinned")

    async def pin_dm_message(self, dm_id: str, message_id: str) -> dict:
        return await self.request("PUT", f"/api/dms/{dm_id}/messages/{message_id}/pin")

    async def unpin_dm_message(self, dm_id: str, message_id: str) -> dict:
        return await self.request("DELETE", f"/api/dms/{dm_id}/messages/{message_id}/pin")

    async def list_dm_pinned(self, dm_id: str) -> list[dict]:
        return await self.request("GET", f"/api/dms/{dm_id}/pinned")

    # ── Moderation ─────────────────────────────────────────────────────

    async def kick_member(self, guild_id: str, user_id: str) -> None:
        await self.request("DELETE", f"/api/servers/{guild_id}/members/{user_id}")

    async def ban_member(self, guild_id: str, user_id: str, reason: str | None = None) -> dict:
        body: dict = {"user_id": user_id}
        if reason is not None:
            body["reason"] = reason
        return await self.request("POST", f"/api/servers/{guild_id}/bans", json=body)

    async def unban_member(self, guild_id: str, user_id: str) -> None:
        await self.request("DELETE", f"/api/servers/{guild_id}/bans/{user_id}")

    async def list_bans(self, guild_id: str) -> list[dict]:
        return await self.request("GET", f"/api/servers/{guild_id}/bans")

    async def set_timeout(
        self, guild_id: str, user_id: str, duration_seconds: int, reason: str | None = None,
    ) -> dict:
        body: dict = {"duration_seconds": duration_seconds}
        if reason is not None:
            body["reason"] = reason
        return await self.request(
            "PUT", f"/api/servers/{guild_id}/members/{user_id}/timeout", json=body,
        )

    async def clear_timeout(self, guild_id: str, user_id: str) -> None:
        await self.request("DELETE", f"/api/servers/{guild_id}/members/{user_id}/timeout")

    async def update_member(
        self, guild_id: str, user_id: str, *, nickname: str | None = None, role: str | None = None,
    ) -> dict:
        body: dict = {}
        if nickname is not None:
            body["nickname"] = nickname
        if role is not None:
            body["role"] = role
        return await self.request(
            "PATCH", f"/api/servers/{guild_id}/members/{user_id}", json=body,
        )

    # ── Roles ──────────────────────────────────────────────────────────

    async def list_roles(self, guild_id: str) -> list[dict]:
        return await self.request("GET", f"/api/servers/{guild_id}/roles")

    async def create_role(
        self, guild_id: str, name: str, *,
        color: str | None = None, permissions: int = 0,
        hoist: bool = False, mentionable: bool = True,
    ) -> dict:
        body: dict = {
            "name": name, "permissions": permissions,
            "hoist": hoist, "mentionable": mentionable,
        }
        if color is not None:
            body["color"] = color
        return await self.request("POST", f"/api/servers/{guild_id}/roles", json=body)

    async def update_role(self, guild_id: str, role_id: str, **fields) -> dict:
        return await self.request(
            "PATCH", f"/api/servers/{guild_id}/roles/{role_id}", json=fields,
        )

    async def delete_role(self, guild_id: str, role_id: str) -> None:
        await self.request("DELETE", f"/api/servers/{guild_id}/roles/{role_id}")

    async def get_member_roles(self, guild_id: str, user_id: str) -> list[str]:
        return await self.request(
            "GET", f"/api/servers/{guild_id}/roles/members/{user_id}",
        )

    async def set_member_roles(
        self, guild_id: str, user_id: str, role_ids: list[str],
    ) -> list[str]:
        return await self.request(
            "PUT", f"/api/servers/{guild_id}/roles/members/{user_id}",
            json={"role_ids": role_ids},
        )

File: hiroo-js/src/http.js
import { HTTPError } from "./errors.js";

/**
 * Thin wrapper over fetch() for the HiRoo REST API.
 *
 * Bot tokens are sent as `Authorization: Bot <token>`; if your token already
 * carries the `Bot ` / `Bearer ` prefix it's used verbatim.
 */
export class HTTPClient {
  constructor({ token, baseURL }) {
    this.token = token;
    this.baseURL = baseURL.replace(/\/$/, "");
  }

  get authHeader() {
    const t = this.token ?? "";
    return /^(bot |bearer )/i.test(t) ? t : `Bot ${t}`;
  }

  async request(method, path, { json, params } = {}) {
    let url = this.baseURL + path;
    if (params) {
      const qs = new URLSearchParams();
      for (const [k, v] of Object.entries(params)) {
        if (v != null) qs.set(k, String(v));
      }
      const q = qs.toString();
      if (q) url += (url.includes("?") ? "&" : "?") + q;
    }
    const init = {
      method,
      headers: {
        "Authorization": this.authHeader,
        "User-Agent": "hiroo-js/0.1",
      },
    };
    if (json !== undefined) {
      init.headers["Content-Type"] = "application/json";
      init.body = JSON.stringify(json);
    }
    const res = await fetch(url, init);
    if (res.status === 204) return null;
    const text = await res.text();
    let body;
    try { body = text ? JSON.parse(text) : null; } catch { body = text; }
    if (!res.ok) throw new HTTPError(res.status, body);
    return body;
  }

  // ── Messages ─────────────────────────────────────────────────────

  async sendMessage({ channelId, dmId, content = "", embeds, components, replyToId }) {
    const body = { content };
    if (embeds) body.embeds = embeds.map((e) => (e?.toJSON ? e.toJSON() : e));
    if (components) body.components = components.map((c) => (c?.toJSON ? c.toJSON() : c));
    if (replyToId) body.reply_to_id = replyToId;
    if (channelId) return this.request("POST", `/api/channels/${channelId}/messages`, { json: body });
    if (dmId) return this.request("POST", `/api/dms/${dmId}/messages`, { json: body });
    throw new Error("Either channelId or dmId required");
  }

  async editMessage(channelId, messageId, content) {
    return this.request("PATCH", `/api/channels/${channelId}/messages/${messageId}`, { json: { content } });
  }

  async deleteMessage(channelId, messageId) {
    return this.request("DELETE", `/api/channels/${channelId}/messages/${messageId}`);
  }

  async addReaction(channelId, messageId, emoji) {
    return this.request("POST", `/api/channels/${channelId}/messages/${messageId}/reactions`, { json: { emoji } });
  }

  // ── Channels ─────────────────────────────────────────────────────

  async getGuildChannels(guildId) {
    return this.request("GET", `/api/servers/${guildId}/channels`);
  }

  /**
   * Create a channel in a server. `parentId` points to a category channel
   * to nest the new channel inside. Types: `text`, `voice`, `announcement`,
   * `category`, `forum`. Requires the bot to have MANAGE_CHANNELS.
   */
  async createChannel(guildId, name, {
    type = "text",
    topic,
    isPrivate = false,
    parentId,
    position = 0,
  } = {}) {
    const body = { name, type, is_private: isPrivate, position };
    if (topic != null) body.topic = topic;
    if (parentId != null) body.parent_id = parentId;
    return this.request("POST", `/api/servers/${guildId}/channels`, { json: body });
  }

  async updateChannel(guildId, channelId, fields) {
    return this.request("PATCH", `/api/servers/${guildId}/channels/${channelId}`, { json: fields });
  }

  async deleteChannel(guildId, channelId) {
    return this.request("DELETE", `/api/servers/${guildId}/channels/${channelId}`);
  }

  async getGuilds() {
    return this.request("GET", "/api/servers");
  }

  // ── Messages (extra) ─────────────────────────────────────────────

  async editDMMessage(dmId, messageId, content) {
    return this.request("PATCH", `/api/dms/${dmId}/messages/${messageId}`, { json: { content } });
  }

  async deleteDMMessage(dmId, messageId) {
    return this.request("DELETE", `/api/dms/${dmId}/messages/${messageId}`);
  }

  async removeReaction(channelId, messageId, emoji) {
    return this.request("DELETE", `/api/channels/${channelId}/messages/${messageId}/reactions`, { params: { emoji } });
  }

  async pinMessage(channelId, messageId) {
    return this.request("PUT", `/api/channels/${channelId}/messages/${messageId}/pin`);
  }

  async unpinMessage(channelId, messageId) {
    return this.request("DELETE", `/api/channels/${channelId}/messages/${messageId}/pin`);
  }

  async listPinned(channelId) {
    return this.request("GET", `/api/channels/${channelId}/messages/pinned`);
  }

  async pinDMMessage(dmId, messageId) {
    return this.request("PUT", `/api/dms/${dmId}/messages/${messageId}/pin`);
  }

  async unpinDMMessage(dmId, messageId) {
    return this.request("DELETE", `/api/dms/${dmId}/messages/${messageId}/pin`);
  }

  async listDMPinned(dmId) {
    return this.request("GET", `/api/dms/${dmId}/pinned`);
  }

  // ── Moderation ─────────────────────────────────────────────────────

  async kickMember(guildId, userId) {
    return this.request("DELETE", `/api/servers/${guildId}/members/${userId}`);
  }

  async banMember(guildId, userId, reason = null) {
    const body = { user_id: userId };
    if (reason != null) body.reason = reason;
    return this.request("POST", `/api/servers/${guildId}/bans`, { json: body });
  }

  async unbanMember(guildId, userId) {
    return this.request("DELETE", `/api/servers/${guildId}/bans/${userId}`);
  }

  async listBans(guildId) {
    return this.request("GET", `/api/servers/${guildId}/bans`);
  }

  async setTimeout(guildId, userId, durationSeconds, reason = null) {
    const body = { duration_seconds: durationSeconds };
    if (reason != null) body.reason = reason;
    return this.request("PUT", `/api/servers/${guildId}/members/${userId}/timeout`, { json: body });
  }

  async clearTimeout(guildId, userId) {
    return this.request("DELETE", `/api/servers/${guildId}/members/${userId}/timeout`);
  }

  async updateMember(guildId, userId, { nickname, role } = {}) {
    const body = {};
    if (nickname !== undefined) body.nickname = nickname;
    if (role !== undefined) body.role = role;
    return this.request("PATCH", `/api/servers/${guildId}/members/${userId}`, { json: body });
  }

  // ── Roles ──────────────────────────────────────────────────────────

  async listRoles(guildId) {
    return this.request("GET", `/api/servers/${guildId}/roles`);
  }

  async createRole(guildId, name, { color, permissions = 0, hoist = false, mentionable = true } = {}) {
    const body = { name, permissions, hoist, mentionable };
    if (color != null) body.color = color;
    return this.request("POST", `/api/servers/${guildId}/roles`, { json: body });
  }

  async updateRole(guildId, roleId, fields) {
    return this.request("PATCH", `/api/servers/${guildId}/roles/${roleId}`, { json: fields });
  }

  async deleteRole(guildId, roleId) {
    return this.request("DELETE", `/api/servers/${guildId}/roles/${roleId}`);
  }

  async getMemberRoles(guildId, userId) {
    return this.request("GET", `/api/servers/${guildId}/roles/members/${userId}`);
  }

  async setMemberRoles(guildId, userId, roleIds) {
    return this.request("PUT", `/api/servers/${guildId}/roles/members/${userId}`, { json: { role_ids: roleIds } });
  }

  // ── Commands (slash) ─────────────────────────────────────────────

  async registerCommand(applicationId, { type = "slash", name, description, options = [], guildId = null }) {
    return this.request("POST", `/api/commands/applications/${applicationId}/commands`, {
      json: { type, name, description, options, guild_id: guildId },
    });
  }

  // ── Interactions ─────────────────────────────────────────────────

  async interactionCallback(interactionId, { type = 4, content = "", embeds, components, ephemeral = false } = {}) {
    const body = { type, content, ephemeral };
    if (embeds) body.embeds = embeds.map((e) => (e?.toJSON ? e.toJSON() : e));
    if (components) body.components = components.map((c) => (c?.toJSON ? c.toJSON() : c));
    return this.request("POST", `/api/interactions/${interactionId}/callback`, { json: body });
  }

  async interactionFollowup(interactionId, { content = "", embeds, components } = {}) {
    const body = { content };
    if (embeds) body.embeds = embeds.map((e) => (e?.toJSON ? e.toJSON() : e));
    if (components) body.components = components.map((c) => (c?.toJSON ? c.toJSON() : c));
    return this.request("POST", `/api/interactions/${interactionId}/followup`, { json: body });
  }
}
